File: frontend/src/api/nodeIpProbes.ts
import apiClient from './client'
import { readAuthToken, readCsrfToken } from '../utils/authToken'

const apiBaseUrl = import.meta.env.VITE_API_BASE_URL || ''

export interface NodeIpProbeItem {
  node_id: number
  status: 'ok' | 'error'
  ip: string | null
  ip_version: 4 | 6 | null
  country_code: string | null
  country_name: string | null
  country_source: string | null
  intelligence_status: 'syncing' | 'pending' | 'enriched' | 'error' | 'disabled' | null
  intelligence_message: string | null
  message: string | null
  probed_at: string
  country_updated_at: string | null
  intelligence_updated_at: string | null
  updated_at: string
}

interface NodeIpProbeListResponse {
  code: string
  data: NodeIpProbeItem[]
}

export type NodeIpProbeEvent =
  | {
      type: 'job_started'
      total_nodes: number
      provider: string
    }
  | {
      type: 'node_completed'
      id: number
      status: 'ok' | 'error'
      ip: string | null
      ip_version: 4 | 6 | null
      country_code: string | null
      country_name: string | null
      message: string | null
      probed_at: string
    }
  | {
      type: 'intelligence_updated'
      id: number
      country_code: string | null
      country_name: string | null
      country_source: string | null
      intelligence_status: NodeIpProbeItem['intelligence_status']
      intelligence_message: string | null
      intelligence_updated_at: string | null
    }
  | {
      type: 'job_completed'
      total_nodes: number
      succeeded: number
      failed: number
      cancelled: boolean
    }

export async function listNodeIpProbes() {
  const { data } = await apiClient.get<NodeIpProbeListResponse>('/api/v1/node-ip-probes')
  return data
}

export async function streamNodeIpProbes(
  ids: number[],
  onEvent: (event: NodeIpProbeEvent) => void,
  signal?: AbortSignal,
) {
  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
    'Cache-Control': 'no-cache',
    Pragma: 'no-cache',
  }
  const token = readAuthToken()
  const csrfToken = readCsrfToken()
  if (token) headers.Authorization = `Bearer ${token}`
  if (csrfToken) headers['X-CSRF-Token'] = csrfToken

  const response = await fetch(`${apiBaseUrl}/api/v1/node-ip-probes/stream`, {
    method: 'POST',
    credentials: 'include',
    headers,
    body: JSON.stringify({ ids }),
    signal,
  })
  if (!response.ok) throw await fetchApiError(response)
  if (!response.body) throw new Error('Streaming response is not available.')

  const reader = response.body.getReader()
  const decoder = new TextDecoder()
  let buffer = ''
  while (true) {
    const { done, value } = await reader.read()
    if (done) break
    buffer += decoder.decode(value, { stream: true })
    const lines = buffer.split('\n')
    buffer = lines.pop() ?? ''
    for (const line of lines) {
      const trimmed = line.trim()
      if (trimmed) onEvent(JSON.parse(trimmed) as NodeIpProbeEvent)
    }
  }
  buffer += decoder.decode()
  if (buffer.trim()) onEvent(JSON.parse(buffer.trim()) as NodeIpProbeEvent)
}

export async function cancelNodeIpProbes() {
  const { data } = await apiClient.post<{ code: string; cancelled: boolean }>('/api/v1/node-ip-probes/cancel')
  return data
}

export interface NodeIpIntelligenceStatus {
  code: string
  enabled: boolean
  base_url: string | null
  source_key: string | null
}

export async function getNodeIpIntelligenceStatus() {
  const { data } = await apiClient.get<NodeIpIntelligenceStatus>('/api/v1/node-ip-probes/intelligence/status')
  return data
}

export async function refreshNodeIpIntelligence(ids: number[]) {
  const { data } = await apiClient.post<{
    code: string
    updated: number
    pending: number
    failed: number
    data: NodeIpProbeItem[]
  }>('/api/v1/node-ip-probes/intelligence/refresh', { ids })
  return data
}

async function fetchApiError(response: Response) {
  let payload: { code?: string; message?: string; error?: string } = {}
  try {
    payload = await response.json()
  } catch {
    payload = {}
  }
  const error = new Error(payload.message ?? payload.error ?? response.statusText) as Error & { code?: string }
  error.code = payload.code
  return error
}

File: backend/src/dto/node_ip_probes.rs
use serde::{Deserialize, Serialize};

use crate::domain::node_ip_probe::NodeIpProbeRecord;

#[derive(Debug, Deserialize)]
pub struct NodeIpProbeRequest {
    pub ids: Vec<i64>,
}

#[derive(Debug, Deserialize)]
pub struct RefreshNodeIpIntelligenceRequest {
    pub ids: Vec<i64>,
}

#[derive(Debug, Deserialize)]
pub struct BatchNodeRiskScoreRequest {
    pub ids: Vec<i64>,
}

#[derive(Debug, Deserialize)]
pub struct UpdateNodeIpCountryRequest {
    pub ip: String,
    pub country_code: String,
    pub country_name: Option<String>,
    pub source: String,
}

#[derive(Debug, Serialize)]
pub struct NodeIpProbeListResponse {
    pub code: &'static str,
    pub data: Vec<NodeIpProbeRecord>,
}

#[derive(Debug, Serialize)]
pub struct NodeIpProbeResponse {
    pub code: &'static str,
    pub data: NodeIpProbeRecord,
}

#[derive(Debug, Serialize)]
pub struct NodeIpIntelligenceStatusResponse {
    pub code: &'static str,
    pub enabled: bool,
    pub base_url: Option<String>,
    pub source_key: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct RefreshNodeIpIntelligenceResponse {
    pub code: &'static str,
    pub updated: usize,
    pub pending: usize,
    pub failed: usize,
    pub data: Vec<NodeIpProbeRecord>,
}

#[derive(Debug, Serialize)]
pub struct BatchNodeRiskScoreResponse {
    pub code: &'static str,
    pub requested: usize,
    pub unique_ips: usize,
    pub complete: usize,
    pub partial: usize,
    pub pending: usize,
    pub failed: usize,
    pub data: Vec<NodeRiskScoreResult>,
}

#[derive(Debug, Serialize)]
pub struct NodeRiskScoreResult {
    pub node_id: i64,
    pub node_name: Option<String>,
    pub ip: Option<String>,
    pub status: &'static str,
    pub intelligence_cache_status: Option<String>,
    pub scamalytics_fraud_score: Option<u8>,
    pub scamalytics_isp_risk_score: Option<u8>,
    pub message: Option<String>,
}

#[derive(Debug, Serialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum NodeIpProbeEvent {
    JobStarted {
        total_nodes: usize,
        provider: &'static str,
    },
    NodeCompleted {
        id: i64,
        status: String,
        ip: Option<String>,
        ip_version: Option<i64>,
        country_code: Option<String>,
        country_name: Option<String>,
        message: Option<String>,
        probed_at: String,
    },
    IntelligenceUpdated {
        id: i64,
        country_code: Option<String>,
        country_name: Option<String>,
        country_source: Option<String>,
        intelligence_status: Option<String>,
        intelligence_message: Option<String>,
        intelligence_updated_at: Option<String>,
        risk_ip: Option<String>,
        risk_status: Option<String>,
        scamalytics_fraud_score: Option<i64>,
        scamalytics_isp_risk_score: Option<i64>,
        risk_checked_at: Option<String>,
        risk_expires_at_unix_ms: Option<i64>,
        risk_message: Option<String>,
    },
    JobCompleted {
        total_nodes: usize,
        succeeded: usize,
        failed: usize,
        cancelled: bool,
    },
}
